test/with_dummyserver/test_socketlevel.py

```diff
@@ -1076,3 +1076,52 @@ def socket_handler(listener):
         self.assertEqual([b'hello, world'], list(r.stream(None)))
 
         done_event.set()
+
+class TestBadContentLength(SocketDummyServerTestCase):
+    def test_strict_content_length(self):
+        done_event = Event()
+
+        def socket_handler(listener):
+            sock = listener.accept()[0]
+
+            buf = b''
+            while not buf.endswith(b'\r\n\r\n'):
+                buf += sock.recv(65536)
+
+            sock.send(
+                b'HTTP/1.1 200 OK\r\n'
+                b'Content-Length: 22\r\n'
+                b'Content-type: text/plain\r\n'
+                b'\r\n'
+                b'hello, world'
+            )
+            done_event.wait(1)
+            sock.close()
+
+        def _make_connection(test_obj):
+            test_obj._start_server(socket_handler)
+            return HTTPConnectionPool(test_obj.host, test_obj.port, maxsize=1)
+
+        def test_get(test_obj):
+            # Test stream read when content length less than headers claim
+            conn = _make_connection(test_obj)
+            get_response = conn.request('GET', url='/', preload_content=False,
+                                        strict_content_length=True)
+            data = get_response.stream(100)
+            # Read "good" data before we try to read again.
+            # This won't trigger till generator is exhausted.
+            next(data)
+            test_obj.assertRaises(ProtocolError, next, data)
+
+        def test_head(test_obj):
+            #Test stream on 0 length body
+            conn = _make_connection(test_obj)
+            head_response = conn.request('HEAD', url='/', preload_content=False,
+                                         strict_content_length=True)
+            data = [chunk for chunk in head_response.stream(1)]
+            assert len(data) == 0
+
+        test_get(self)
+        test_head(self)
```

Comment:
I'd definitely like to do that, I wasn't sure if there was a way to reuse the socket_handler without having to rewrite it for each test? I can if needed, I just didn't want to add redundant bloat if it wasn't needed.